poc_iot_injector/src/receipt_txn.rs

```diff
@@ -0,0 +1,244 @@
+};
+use rust_decimal::{prelude::ToPrimitive, Decimal, MathematicalOps};
+use rust_decimal_macros::dec;
+use sha2::{Digest, Sha256};
+
+// TODO: These should ideally be coming in from some configuration service
+const WITNESS_REDUNDANCY: u32 = 4;
+const POC_REWARD_DECAY_RATE: Decimal = dec!(0.8);
+const HIP15_TX_REWARD_UNIT_CAP: Decimal = dec!(2.0);
+
+// TODO: This should be done better?
+const MHZTOHZ: u32 = 1000000;
+
+type PocPath = Vec<BlockchainPocPathElementV1>;
+
+pub fn handle_report_msg(
+    msg: prost::bytes::BytesMut,
+    keypair: &Keypair,
+    timestamp: i64,
+) -> Result<Option<(BlockchainTxnPocReceiptsV2, Vec<u8>, String)>> {
+    // NOTE: This is intentionally not Vec::with_capacity(1) since we _may_ add multi-hop PocPath
+    // in the future
+    let mut path: PocPath = vec![];
+
+    let LoraValidPocV1 {
+        poc_id: _poc_id,
+        beacon_report,
+        witness_reports,
+    } = LoraValidPocV1::decode(msg)?;
+
+    let poc_witnesses = construct_poc_witnesses(witness_reports)?;
+
+    // We expect poc_receipt to always be here, once it is in a report
+    if let Some(poc_receipt) = construct_poc_receipt(beacon_report, poc_witnesses.len() as u32)? {
+        // TODO: Double check whether the gateway in the poc_receipt is challengee?
+        let path_element =
+            construct_path_element(poc_receipt.clone().gateway, poc_receipt, poc_witnesses)?;
+
+        path.push(path_element);
+
+        Ok(Some(construct_txn(path, timestamp, keypair)?))
+    } else {
+        Ok(None)
+    }
+}
+
+fn construct_txn(
+    path: PocPath,
+    timestamp: i64,
+    keypair: &Keypair,
+) -> Result<(BlockchainTxnPocReceiptsV2, Vec<u8>, String)> {
+    let mut txn = BlockchainTxnPocReceiptsV2 {
+        challenger: keypair.public_key().to_vec(),
+        secret: vec![],
+        onion_key_hash: vec![],
+        path,
+        fee: 0,
+        signature: vec![],
+        block_hash: vec![],
+        timestamp: timestamp as u64,
+    };
+
+    txn.signature = sign_txn(&txn, keypair)?;
+    let (txn_hash, txn_hash_b64url) = hash_txn(&txn);
+    Ok((txn, txn_hash, txn_hash_b64url))
+}
+
+fn construct_path_element(
+    challengee: Vec<u8>,
+    poc_receipt: BlockchainPocReceiptV1,
+    poc_witnesses: Vec<BlockchainPocWitnessV1>,
+) -> Result<BlockchainPocPathElementV1> {
+    let path_element = BlockchainPocPathElementV1 {
+        challengee,
+        receipt: Some(poc_receipt),
+        witnesses: poc_witnesses,
+    };
+    Ok(path_element)
+}
+
+fn construct_poc_witnesses(
+    witness_reports: Vec<LoraValidWitnessReportV1>,
+) -> Result<Vec<BlockchainPocWitnessV1>> {
+    let num_witnesses = witness_reports.len();
+    let mut poc_witnesses: Vec<BlockchainPocWitnessV1> = Vec::with_capacity(num_witnesses);
+    for witness_report in witness_reports {
+        let LoraValidWitnessReportV1 {
+            received_timestamp: _witness_received_timestamp,
+            location: _witness_location,
```

Comment:
do you need these if you're not using them? can you not match on only the fields you care about and ignore the rest with `..`?

Reply:
right, forgot about that. Updated 9cf0023